inc/az_ulib_ustream_forward_base.h

```diff
@@ -0,0 +1,374 @@
+ *          can copy, the read shall fail with pre-condition.
+ *      - If the provided interface is `NULL`, the read shall fail with precondition.
+ *      - If the provided interface is not the implemented ustream_forward type, the read shall fail with
+ *          pre-condition.
+ *      - If the provided local buffer is `NULL`, the read shall fail with precondition.
+ *      - If the provided return size pointer is `NULL`, the read shall fail with precondition.
+ *
+ * @param[in]       ustream_forward_instance    The #az_ulib_ustream_forward* with the interface of the ustream_forward.
+ * @param[out]      buffer              The `uint8_t* const` that points to the local buffer.
+ * @param[in]       buffer_length       The `size_t` with the size of the local buffer.
+ * @param[out]      size                The `size_t* const` that points to the place where the
+ *                                      read shall store the number of valid `uint8_t` values
+ *                                      returned in the local buffer.
+ *
+ * @pre     \p ustream_forward_instance shall not be `NULL`.
+ * @pre     \p ustream_forward_instance shall be a valid ustream_forward that is the implemented ustream_forward type.
+ * @pre     \p buffer shall not be `NULL`.
+ * @pre     \p buffer_length shall not be bigger than 0.
+ * @pre     \p size shall not be `NULL`.
+ *
+ * @return The #az_result with the result of the read operation.
+ *      @retval #AZ_OK                        If the ustream_forward copied the content of the `Data
+ *                                            Source` to the local buffer with success.
+ *      @retval #AZ_ERROR_ULIB_BUSY           If the resource necessary to read the ustream_forward
+ *                                            content is busy.
+ *      @retval #AZ_ERROR_CANCELED            If the read of the content was cancelled.
+ *      @retval #AZ_ULIB_EOF                  If there are no more `uint8_t` values in the `Data
+ *                                            Source` to read.
+ *      @retval #AZ_ERROR_NOT_ENOUGH_SPACE    If there is not enough memory to execute the read.
+ *      @retval #AZ_ERROR_ULIB_SECURITY       If the read was denied for security reasons.
+ *      @retval #AZ_ERROR_ULIB_SYSTEM         If the read operation failed on the system level.
+ */
+AZ_INLINE az_result az_ulib_ustream_forward_read(
+    az_ulib_ustream_forward* ustream_forward_instance,
+    uint8_t* const buffer,
+    size_t buffer_length,
+    size_t* const size)
+{
+  return ustream_forward_instance->control_block->api->read(ustream_forward_instance, buffer, buffer_length, size);
+}
+
+/**
+ * @brief   Returns the remaining size of the ustream_forward.
+ *
+ *  This API returns the number of bytes between the current position and the end of the ustream_forward.
+ *
+ *  The `az_ulib_ustream_forward_get_remaining_size` API shall follow the following minimum requirements:
```

Comment:
>az_ulib_ustream_forward_get_remaining_size [](http://example.com/codeflow?start=9&length=42)

Thinking more about this interface, we should not expose this API, We should return only the length of the stream.

Reply:
I think if that's the case then we may want to do away with `read` altogether. let's discuss tomorrow

Reply:
now I see your point. agreed.